adapter/config/ingress/conversion.go

```diff
@@ -0,0 +1,173 @@
+			out[model.IngressRuleDescriptor.Key(ingressRule)] = ingressRule
+		}
+	}
+
+	return out
+}
+
+func createIngressRule(name, host, path, namespace, domainSuffix string,
+	backend v1beta1.IngressBackend, tlsSecret string) *proxyconfig.IngressRule {
+	rule := &proxyconfig.IngressRule{
+		Name:        name,
+		Destination: serviceHostname(backend.ServiceName, namespace, domainSuffix),
+		TlsSecret:   tlsSecret,
+		Match: &proxyconfig.MatchCondition{
+			HttpHeaders: make(map[string]*proxyconfig.StringMatch, 2),
+		},
+	}
+	switch backend.ServicePort.Type {
+	case intstr.Int:
+		rule.DestinationServicePort = &proxyconfig.IngressRule_DestinationPort{
+			DestinationPort: int32(backend.ServicePort.IntValue()),
+		}
+	case intstr.String:
+		rule.DestinationServicePort = &proxyconfig.IngressRule_DestinationPortName{
+			DestinationPortName: backend.ServicePort.String(),
+		}
+	}
+
+	if host != "" {
+		rule.Match.HttpHeaders[model.HeaderAuthority] = &proxyconfig.StringMatch{
+			MatchType: &proxyconfig.StringMatch_Exact{Exact: host},
+		}
+	}
+
+	if path != "" {
+		if isRegularExpression(path) {
+			if strings.HasSuffix(path, ".*") && !isRegularExpression(strings.TrimSuffix(path, ".*")) {
+				rule.Match.HttpHeaders[model.HeaderURI] = &proxyconfig.StringMatch{
+					MatchType: &proxyconfig.StringMatch_Prefix{Prefix: strings.TrimSuffix(path, ".*")},
+				}
+			} else {
+				rule.Match.HttpHeaders[model.HeaderURI] = &proxyconfig.StringMatch{
+					MatchType: &proxyconfig.StringMatch_Regex{Regex: path},
+				}
+			}
+		} else {
+			rule.Match.HttpHeaders[model.HeaderURI] = &proxyconfig.StringMatch{
+				MatchType: &proxyconfig.StringMatch_Exact{Exact: path},
+			}
+		}
+	}
+
+	return rule
+}
+
+// encodeIngressRuleName encodes an ingress rule name for a given ingress resource name,
+// as well as the position of the rule and path specified within it, counting from 1.
+// ruleNum == pathNum == 0 indicates the default backend specified for an ingress.
+func encodeIngressRuleName(ingressName, ingressNamespace string, ruleNum, pathNum int) string {
+	return fmt.Sprintf("%s.%s.%d.%d", ingressNamespace, ingressName, ruleNum, pathNum)
+}
+
+// decodeIngressRuleName decodes an ingress rule name previously encoded with encodeIngressRuleName.
+func decodeIngressRuleName(name string) (ingressName, ingressNamespace string, ruleNum, pathNum int, err error) {
+	parts := strings.Split(name, ".")
+	if len(parts) != 4 {
+		err = fmt.Errorf("could not decode string into ingress rule name: %s", name)
+		return
+	}
+
+	ingressNamespace = parts[0]
+	ingressName = parts[1]
+	ruleNum, ruleErr := strconv.Atoi(parts[2])
+	pathNum, pathErr := strconv.Atoi(parts[3])
+
+	if pathErr != nil || ruleErr != nil {
+		err = multierror.Append(
+			fmt.Errorf("could not decode string into ingress rule name: %s", name),
+			pathErr, ruleErr)
+		return
+	}
+
+	return
+}
+
+// isRegularExpression determines whether the given string s is a non-trivial regular expression,
```

Comment:
Warn that Golang regular expression might not be a 1-1 match with this regular expression?